src/CiviEntityStorage.php

```diff
@@ -362,13 +365,15 @@ protected function initFieldValues(ContentEntityInterface $entity, array $values
           // Handle if the value provided is a timestamp.
           // @note: This only occurred during test migrations.
           elseif (is_numeric($item[$main_property_name])) {
-            $item_values[$delta][$main_property_name] = (new \DateTime())->setTimestamp($item[$main_property_name])->format(DATETIME_DATETIME_STORAGE_FORMAT);
+            $item_values[$delta][$main_property_name] = (new \DateTime('now', new \DateTimeZone('UTC')))->setTimestamp($item[$main_property_name])->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT);
           }
           else {
+            $datetime_format = $definition->getSetting('datetime_type') === DateTimeItem::DATETIME_TYPE_DATE ? DateTimeItemInterface::DATE_STORAGE_FORMAT : DateTimeItemInterface::DATETIME_STORAGE_FORMAT;
             // CiviCRM gives us the datetime in the users timezone (or no
             // timezone at all) but Drupal expects it in UTC. So, we need to
             // convert from the users timezone into UTC.
-            $item_values[$delta][$main_property_name] = (new \DateTime($item[$main_property_name], new \DateTimeZone(drupal_get_user_timezone())))->setTimezone(new \DateTimeZone('UTC'))->format(DATETIME_DATETIME_STORAGE_FORMAT);
+            $datetime_value = (new \DateTime($item[$main_property_name], new \DateTimeZone('UTC')))->format($datetime_format);
```

Comment:
I haven't tested this, but I suspect the changes here to UTC are not quite right. CiviCRM doesn't keep the datetimes in UTC time, it keeps them in local time. So, a lot of the unusual date formulations here are converting the time in the database from local time into UTC time, because Drupal expects them to be in UTC.

@mglaman Have you tested this with other timezones?

Reply:
I have not. But I realized there was proof of something weird in the test assertions.

https://github.com/eileenmcnaughton/civicrm_entity/pull/190/files#diff-22cf275b4dbef73100eb9fe7e1347fb8L44

```
-    $this->assertEquals('2018-05-02T07:00:00', $entity->get('start_date')->value);
+    $this->assertEquals('2018-05-02T17:00:00', $entity->get('start_date')->value);
```

The test data has the value 

```
        'start_date' => '2018-05-02 17:00:00',
        'event_start_date' => '2018-05-02 17:00:00',
        'end_date' => '2018-05-04 17:00:00',
        'event_end_date' => '2018-05-04 17:00:00',
```

The original test was showing a 10-hour difference.  With the adjustment, it's now correct. Drupal also executes it's PHPUnit with the timezone set to Sydney, Australia

```php
// Set the default timezone. While this doesn't cause any tests to fail, PHP
// complains if 'date.timezone' is not set in php.ini. The Australia/Sydney
// timezone is chosen so all tests are run using an edge case scenario (UTC+10
// and DST). This choice is made to prevent timezone related regressions and
// reduce the fragility of the testing system in general.
date_default_timezone_set('Australia/Sydney');
```

It's also more concrete than trusting `drupal_get_user_timezone` which is per-user. Maybe we need to check the timezone setting in CiviCRM instead? Because that's more reliable than Drupal's configured timezone, technically.

Reply:
The reason I think it's using the user timezone, is that we want the timezone to _appear_ to be unprocessed. So, if an event has '2020-01-01T10:00:00' in the database, we want Drupal to show it to the user as exactly January 1st, 2020 @ 10am, no matter what the user has their timezone set at.

So, that mess of original code is basically saying, "Let's interpret the database datetime as happening in the current user's timezone, and then turn it into UTC for Drupal, which will later turn it BACK into the user's timezone, so it's displayed exactly as it is in the database." Unfortunately, while super weird, I think this is actually correct. We should have some tests for this.

Reply:
I ran into a timezone issue with D8/CiviCRM myself recently. Just D8/CiviCRM native -> Contributions entered are coming in (have received dates) anywhere between user time zone (from D8) and UTC: https://civicrm.stackexchange.com/questions/33541/timezone-issue-authenticated-vs-anonymous-a-6h-time-warp